source/dispatch/functor_map.js

```diff
@@ -0,0 +1,13 @@
+import _isArrayLike from '../internal/_isArrayLike';
+import _isFunction from '../internal/_isFunction';
+import _isObject from '../internal/_isObject';
+import array_map from '../instances/array_map';
+import function_map from '../instances/function_map';
+import object_map from '../instances/object_map';
+
+export default function functor_map(fn, functor) {
+  if (_isArrayLike(functor)) return array_map(fn, functor);
```

Comment:
I have just `_isArray` below for `concat`. Both should be `isArray` probably, but we used to just fall through to array implementation for `map`, so this is more like the old behaviour?

Reply:
Does that apply to `Array` too?  How about `Function` or `Object`?  The point here is to decide where we want to apply it *beyond* any dispatching.

Reply:
Taking an Arguments object and returning an array is not compliant with the spec:

> `map` must return a value of the same Functor

If we want to be able to map over Arguments objects we'd need to support this:

```haskell
map :: (a -> b) -> Arguments a -> Arguments b
```

This seems silly to me, but I wouldn't oppose it.

I think we should provide the same implementations for built-in types `Z.map` does.

Reply:
This comes back to our fundamental disagreement on Ramda's relationship with FantasyLand.

I would like Ramda to be able to ergonomically interoperate with many tools: things like Immutable.js, streaming libraries, reactive toolkits, and types derived from specifications such as FantasyLand.

I don't think of any of those specifying how Ramda should work.

So I don't think of FantasyLand as a specification for Ramda, only as a way to specify certain types.  I would like us to be able to interoperate with those types, but that does not imply that our functions can't do things incompatible with FL laws for other types.

I don't have any strong desire to work with `Arguments` or `NodeList` objects in `map`, but I don't think of FantasyLand as specifying that we cannot do this if we choose:

```js
const squareArgs = function() {return map(n => n * n, arguments)}
squareArgs(1, 2, 3, 4) //=> [1, 4, 9, 16]
```

even though that invocation of `map` does not conform to the FL spec.

Reply:
I think I now see your point, Scott. Are you suggesting that we behave lawfully for values which do satisfy a particular Fantasy Land type class, but that we're free to define ad hoc behaviour—rather than throwing—for values which do not?

Reply:
I think those types that can satisfy normal algebraic laws -- and whose functions we supply -- should by all means supply those laws.  But I think that we should not ever try to enforce those laws on our own, and I believe that there are common usages around some of these function names, regardless of whether the types obey the laws as specified by FantasyLand.  I would like to be able to interoperate with this types nonetheless.

For the same reason I would like to be able to do `map(square, Just(5))`, even though `Maybe` supplies a method and I can so `Just(5).map(square)` instead, I would also like to be able do `map(square, myStream)` even when `myStream.map(square)` is available.  And this is true, even if someone hasn't proven the composability of `map` on these streams.  It's the same abstraction and should be offered in the same way.

So yes, I would say behaving lawfully where we can is important, but R.map should not be the function that tries to serve all lawful functors and keep out anything that isn't.

Reply:
> It's the same abstraction and should be offered in the same way.

Ah, I see the distinction you're making. You're happy with `map(square, myStream)` but you'd consider `empty($divs)` an abuse? (The jQuery expression `$divs.empty()` has nothing to do with the Monoid type class.)

Reply:
Yes, and that's a _caveat emptor_, of course.  If you use this with types that don't conform, bad things could happen.  But from Ramda's perspective that's no different from

```js
 map(square, {
  'fantasy-land/map': launchTheMissiles
});
```

We don't get to test that the functions specified by the types conform to any particular rules, so let's not bother restricting the dispatch to those claiming to.